Refactor logic for checking storage add-on availability

fredrikekelund opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/wp-calypso/issues/97225

## Proposed Changes

The main bug this PR fixes is that free sites with the 100 GB add-on keep displaying the 50 GB add-on on `/add-ons/:site_slug`—indicating that it's available for purchase even though it's not. I suspect this is also true for Premium sites, but I haven't tested it.

While working on this issue, I found the existing availability logic for storage add-ons to be quite spaghetti-esque, so this PR also takes a stab at simplifying that logic. I've implemented the following refactorings:

1. Remove the `storage-enabled` feature flag. It has been enabled in all environments since Jun, 2023 (https://github.com/Automattic/wp-calypso/pull/78775)
2. Reduce props drilling between `AddOnsMain`, `AddOnsGrid`, and `AddOnCard`
3. Remove add-on availability checks from `useAddOns`. Why? Because we previously combined other hooks like `useAddOnPurchaseStatus` with the availability checks in `useAddOns`. With my refactoring, we clarify the purpose of `useAddOns` solely as a data provider.
4. Added a `useStorageAddOnAvailability` hook for checking if a single storage add-on is available for purchase when compared to the maximum site storage offered on WordPress.com.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I've tagged a few different people for review. @taipeicoder, if you have a moment to review this, I'd appreciate it, since you worked on this logic previously. @rcrdortiz, if you have some groundskeeping time left this afternoon, I'd also appreciate a review. 

1. Apply 168696-ghe-Automattic/wpcom to your sandbox
2. Sandbox `public-api.wordpress.com`
3. Do NOT test with the store sandbox enabled. This [alters](https://github.com/Automattic/wp-calypso/issues/97225#issuecomment-2539354841) your results.
4. I've included a table with four different site configurations to test on three different pages. The slugs for those pages are:
 3.1. Add-ons page: `/add-ons/:site_slug`
 3.2. Site overview page: `/sites/overview/:site_slug`

|                                              | **Add-ons page: 50 GB add-on**      | **Add-ons page: 100 GB add-on**     | **Storage meter on site overview page** |
| -------------------------------------------- | ----------------------------------- | ----------------------------------- | --------------------------------------- |
| **Paid site with 100 GB storage add-on** | Hidden                              | Visible with “Manage add-on” button | No “Need more storage?” link            |
| **Paid site with 50 GB storage add-on**  | Visible with “Manage add-on” button | Purchasable                         | “Need more storage?” link               |
| **Free site without storage add-ons**        | Purchasable                         | Purchasable                         | “Need more storage?” link               |
| **Free site with 100 GB storage add-on** | Hidden                              | Visible with “Manage add-on” button | No “Need more storage?” link            |


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

fredrikekelund committed:
WIP - Refactor availability logic for storage add-ons

fredrikekelund committed:
Simplify

fredrikekelund committed:
Simplify

fredrikekelund committed:
Fix regression

fredrikekelund committed:
Fix type errors

fredrikekelund committed:
Take new API data into account

fredrikekelund commented:
@southp, I worked several rounds on this before successfully replicating the issue from your issue https://github.com/Automattic/wp-calypso/issues/97225

I'd appreciate a review when you have a moment 🙏

southp commented:
Thanks for working on this, @fredrikekelund . I will prioritize reviewing this. Sorry that I deployed overlapped clean-up before noticing your work though:

* https://github.com/Automattic/wp-calypso/pull/97391
* https://github.com/Automattic/wp-calypso/pull/97394

Hopefully I won't cause you too much code conflict troubles.

taipeicoder reviewed (approved):
Tested the scenarios listed in the table from the Testing Instructions section, and they work as expected 👍

southp commented on packages/data-stores/src/add-ons/hooks/use-add-on-purchase-status.ts:
```diff
@@ -1,17 +1,23 @@
 
+type AddOnPurchaseStatus = {
+	available: boolean;
+	hidden?: boolean;
```
I don't see this flag used anywhere. I suppose we can remove it :)

fredrikekelund replied:
Good eyes! This is a leftover from a solution I pursued previously.

southp commented on packages/data-stores/src/add-ons/hooks/use-available-storage-add-ons.ts:
```diff
@@ -1,38 +1,50 @@
+	storage?: SiteMediaStorage
+): boolean {
+	if ( ! storage ) {
+		return true;
```
Could you help me better understand the reason of returning `true` here? Without context, my interpretation is that this hook would return `true` when the app hasn't or can't provide it with the correct media storage data, which sounds suspicious.

fredrikekelund replied:
Good question. `storage` is undefined while the `/media-storage` API endpoint is loading. I agree it makes more sense to handle the loading state elsewhere 👍

southp commented on packages/data-stores/src/add-ons/hooks/use-available-storage-add-ons.ts:
```diff
@@ -1,38 +1,50 @@
+	const maxStorageExcludingAddons = storage.maxStorageBytesExcludingAddons / Math.pow( 1024, 3 );
+	const existingAddOnStorage = currentMaxStorage - maxStorageExcludingAddons;
+
+	return existingAddOnStorage < quantity;
```
This should work most of the time here. However, I'd like to point out that expecting JS division to always produce integral values is in general not recommended, since they are double-precision floats underneath. 

Also, if the sole usage of `maxStorageExcludingAddons` introduced in D168645-code is for doing this computation, could we simply return `existingAddOnStorage` from the endpoint instead? That would eliminate the floating number computation here all together as well.

fredrikekelund replied:
I'm not sure how big of a problem this is in practice, but I take your point that we might as well return `existingAddOnStorage` directly from the API endpoint, since that's what we ultimately want anyway. I've updated the back-end PR (now at 168696-gh-Automattic/wpcom).

Let me know your thoughts!

southp commented on packages/data-stores/src/add-ons/hooks/use-available-storage-add-ons.ts:
```diff
@@ -1,38 +1,50 @@
-	}, [ availableStorageUpgrade, storageAddOns ] );
+		return storageAddOns
+			.filter( ( addOn ) => addOn !== null )
+			.filter( ( addOn ): addOn is AddOnMeta =>
```
`storageAddOns` should be always in the type of `[AddOnMeta]` here, so filtering on the dynamic type here won't be necessary. If not, we should fix `useStorageAddOns` instead to fully leverage the power of type checking of TS.

fredrikekelund replied:
My local TS didn't complain without `addOn is AddOnMeta`, but the CI process did, which is why I added this in 3a40d561f59300ba0101ca4687ba6856121f6496

Typescript has (or has had) trouble with inferring types in filtered arrays. So, I'd happily remove this, but I don't think we can without upsetting the CI jobs.

southp replied:
That's unfortunate. In that case, let's leave it there for now.

southp commented on packages/data-stores/src/add-ons/hooks/use-storage-add-on-availability.ts:
```diff
@@ -0,0 +1,25 @@
+	const mediaStorage = useSiteMediaStorage( { siteIdOrSlug: selectedSiteId } );
+
+	if ( addOnMeta.productSlug !== PRODUCT_1GB_SPACE ) {
+		return true;
```
I'm wondering if returning `true` here is correct, since that means the passed in `addOnMeta` doesn't belong to a storage add-on, but the hook considers it as an available storage add-on. Should it return false instead?

fredrikekelund replied:
Tricky. I pondered this for a bit and landed on returning an enum. Let me know what you think, @southp

southp replied:
It's better now. I have a feeling that we can further generalize and simplify it as a general add-on level concept so we don't have to treat storage add-on as a special case here. However, let's not hold ourselves back and further bloat this PR :)

southp commented on packages/data-stores/src/add-ons/hooks/use-add-on-purchase-status.ts:
```diff
@@ -1,17 +1,23 @@
+type AddOnPurchaseStatus = {
+	available: boolean;
+	hidden?: boolean;
+	text?: ReturnType< typeof i18n.translate >;
```
A side note that I think the user-facing text we display shouldn't have been part of this hook. The text is part of UI, so should be derived from the calypso side based on the status value. 

It's not related to this PR, though; just wanted to raise it :)

fredrikekelund replied:
Good point. This is already a meaty PR, though, so I'll leave it for another one

fredrikekelund committed:
Merge branch 'trunk' into f26d/refactor-storage-addon-availability

fredrikekelund committed:
Fix merge

fredrikekelund committed:
Address review comments

fredrikekelund committed:
Return enum from useStorageAddOnAvailability

fredrikekelund committed:
Appease TS

fredrikekelund committed:
Ensure storage quantity doesn't exceed STORAGE_LIMIT

fredrikekelund committed:
Adapt front-end to back-end

fredrikekelund commented:
> I deployed overlapped clean-up before noticing your work though

No worries, the conflicts weren't too bad.

I've addressed your comments, @southp—this should be good for another look now!

southp reviewed (approved):
Thanks for addressing all of my comments. I really appreciate your effort not only fixing the bug but also making the code better in a balanced way 🙇🏼

fredrikekelund committed:
Merge branch 'trunk' into f26d/refactor-storage-addon-availability
